enhancements/cluster-logging/forwarder-label-selector.md

```diff
@@ -0,0 +1,193 @@
+
+```go
+// A label selector is a label query over a set of resources. The result of matchLabels and
+// matchExpressions are ANDed. An empty label selector matches all objects. A null
+// label selector matches no objects.
+type LabelSelector struct {
+	// matchLabels is a map of {key,value} pairs. A single {key,value} in the matchLabels
+	// map is equivalent to an element of matchExpressions, whose key field is "key", the
+	// operator is "In", and the values array contains only "value". The requirements are ANDed.
+	// +optional
+	MatchLabels map[string]string `json:"matchLabels,omitempty" protobuf:"bytes,1,rep,name=matchLabels"`
+	// matchExpressions is a list of label selector requirements. The requirements are ANDed.
+	// +optional
+	MatchExpressions []LabelSelectorRequirement `json:"matchExpressions,omitempty" protobuf:"bytes,2,rep,name=matchExpressions"`
+}
+```
+
+### User Stories
+
+#### Select logs using simple equality-based selector
+
+For exmaple, forward logs with labels `environment=product` and `app=nginx`:
+
+
+```yaml
+apiVersion: "logging.openshift.io/v1"
+kind: "ClusterLogForwarder"
+metadata:
+  name: "instance"
+spec:
+  pipelines:
+    - inputRefs: [ myLogs ]
+      outputRefs: [ default  ]
+  inputs:
+    - name: myLogs
+      application:
+        selector:
+          matchLabels:
+            environment: production
+            app: nginx
+```
+
+#### Select logs using both equality-based and set-based selectors
+
+For example, forward logs with label `component=redis` where label `tier` has
+value `cache` or `proxy` and label `environment` is not `dev`
+
+
+```yaml
+ppapiVersion: "logging.openshift.io/v1"
+kind: "ClusterLogForwarder"
+metadata:
+  name: "instance"
+spec:
+  pipelines:
+    - inputRefs: [ myLogs ]
+      outputRefs: [ default  ]
+  inputs:
+    - name: myLogs
+      application:
+        selector:
+          matchLabels:
+            component: redis
+          matchExpressions:
+            - {key: tier, operator: In, values: [cache, proxy]}
+            - {key: environment, operator: NotIn, values: [dev]}
+```
+
+#### Combine namespace and label selectors
+
+For example: forward logs from namespaces `app1` or `app2` with label
+`environment=production`
+
+
+```yaml
+apiVersion: "logging.openshift.io/v1"
+kind: "ClusterLogForwarder"
+metadata:
+  name: "instance"
+spec:
+  pipelines:
+    - inputRefs: [ myLogs ]
+      outputRefs: [ default  ]
+  inputs:
+    - name: myLogs
+      application:
+        selector:
+          matchLabels:
+            environment: production
+        namespaces: 
+        - app1
+        - app2
+```
+
+### Implementation Details/Notes/Constraints [optional]
+
+The implementation will be similar to the existing `namespaces` selector, but
+with some key differences:
+- Query the k8s API to find IDs of all Pods that match the label selector.
```

Comment:
This would be an optimization as regardless if implementation this translates to either:

* some from of include/exclude mechansim to collect or not
* collect everything and then filter.

Note we need to also be considerate of churning on cycling the collector/normalizer as i image pod changes are frequent and for efficiency we likely dont want to bounce collectors on every pod change

Reply:
Excellent point, re-wrote the implementation section.